app/scripts/controllers/overview.js

```diff
@@ -498,6 +538,173 @@ angular.module('openshiftConsole')
 
         $scope.$on('$destroy', function(){
           DataService.unwatchAll(watches);
+          $(window).off('click.topologyItem');
+          $('.kube-topology g').popover('destroy');
         });
+
+        // Topology view specific code
+        var updateTimeout = null;
+
+        function updateTopology() {
+          updateTimeout = null;
+
+          var topologyRelations = [];
+          var topologyItems = { };
+
+          // Because metadata.uid is not unique among resources
+          function makeId(resource) {
+            return resource.kind + resource.metadata.uid;
+          }
+
+          // Add the services
+          angular.forEach($scope.services, function(service) {
+            topologyItems[makeId(service)] = service;
+          });
+
+          var isRecentDeployment = $filter('isRecentDeployment');
+          var isVisibleDeployment = function(deployment) {
+            // If this is a replication controller and not a deployment, then it's visible.
+            var dcName = $filter('annotation')(deployment, 'deploymentConfig');
+            if (!dcName) {
+              return true;
+            }
+
+            // If the deployment is active, it's visible.
+            if ($filter('hashSize')($scope.podsByDeployment[deployment.metadata.name]) > 0) {
+              return true;
+            }
+
+            // Wait for deployment configs to load.
+            if (!$scope.deploymentConfigs) {
+              return false;
+            }
+
+            // If the deployment config has been deleted and the deployment has no replicas, hide it.
+            // Otherwise all old deployments for a deleted deployment config will be visible.
+            var dc = $scope.deploymentConfigs[dcName];
+            if (!dc) {
+              return false;
+            }
+
+            // Show the deployment if it's recent (latest or in progress) or if it's scalable.
+            return isRecentDeployment(deployment, dc) || $scope.isScalableDeployment(deployment);
+          };
+
+          // Add everything related to services, each of these tables are in
```

Comment:
It may also be nice/easier to maintain if we broke out each of these next sets of `[].forEach` loops into separate named functions.